File: data/derived_cot/rq1_traces/translation/glm5.1/java_to_cpp/code_output/VectorUtil.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <cmath>
#include <stdexcept>

class VectorUtil {
public:
    static double similarity(const std::vector<double>& vector1, const std::vector<double>& vector2) {
        double dotProduct = 0.0;
        double norm1 = 0.0;
        double norm2 = 0.0;
        for (size_t i = 0; i < vector1.size(); i++) {
            dotProduct += vector1[i] * vector2[i];
            norm1 += std::pow(vector1[i], 2);
            norm2 += std::pow(vector2[i], 2);
        }
        double denominator = std::sqrt(norm1) * std::sqrt(norm2);
        return denominator == 0 ? 0.0 : dotProduct / denominator;
    }

    static std::vector<double> cosineSimilarities(const std::vector<double>& vector1, const std::vector<std::vector<double>>& vectorsAll) {
        std::vector<double> similarities;
        double norm1 = std::sqrt(dotProduct(vector1, vector1));
        if (norm1 == 0) {
            for (size_t i = 0; i < vectorsAll.size(); i++) {
                similarities.push_back(0.0);
            }
            return similarities;
        }
        for (const auto& vector2 : vectorsAll) {
            double norm2 = std::sqrt(dotProduct(vector2, vector2));
            if (norm2 == 0) {
                similarities.push_back(0.0);
            } else {
                double sim = dotProduct(vector1, vector2) / (norm1 * norm2);
                similarities.push_back(sim);
            }
        }
        return similarities;
    }

    static double nSimilarity(const std::vector<std::vector<double>>& vectorList1, const std::vector<std::vector<double>>& vectorList2) {
        if (vectorList1.empty() || vectorList2.empty()) {
            throw std::invalid_argument("At least one of the passed lists is empty.");
        }
        std::vector<double> avgVector1 = averageVector(vectorList1);
        std::vector<double> avgVector2 = averageVector(vectorList2);
        return similarity(avgVector1, avgVector2);
    }

    static std::unordered_map<std::string, double> computeIdfWeightDict(int totalNum, const std::unordered_map<std::string, double>& numberDict) {
        std::unordered_map<std::string, double> result;
        for (const auto& entry : numberDict) {
            double count = entry.second;
            double weight = std::log((totalNum + 1) / (count + 1));
            result[entry.first] = weight;
        }
        return result;
    }

private:
    static double dotProduct(const std::vector<double>& vector1, const std::vector<double>& vector2) {
        double result = 0.0;
        for (size_t i = 0; i < vector1.size(); i++) {
            result += vector1[i] * vector2[i];
        }
        return result;
    }

    static std::vector<double> averageVector(const std::vector<std::vector<double>>& vectors) {
        std::vector<double> avgVector(vectors[0].size());
        for (const auto& vector : vectors) {
            for (size_t i = 0; i < vector.size(); i++) {
                avgVector[i] += vector[i];
            }
        }
        for (size_t i = 0; i < avgVector.size(); i++) {
            avgVector[i] /= vectors.size();
        }
        return avgVector;
    }
};

File: data/derived_cot/rq1_traces/translation/glm5.1/java_to_py/code_output/VectorUtil.py
import math
from typing import Dict, List


class VectorUtil:

    @staticmethod
    def similarity(vector1: List[float], vector2: List[float]) -> float:
        dot_product = 0.0
        norm1 = 0.0
        norm2 = 0.0
        for i in range(len(vector1)):
            dot_product += vector1[i] * vector2[i]
            norm1 += vector1[i] ** 2
            norm2 += vector2[i] ** 2
        denominator = math.sqrt(norm1) * math.sqrt(norm2)
        return 0.0 if denominator == 0 else dot_product / denominator

    @staticmethod
    def cosineSimilarities(vector1: List[float], vectors_all: List[List[float]]) -> List[float]:
        similarities: List[float] = []
        norm1 = math.sqrt(VectorUtil._dotProduct(vector1, vector1))
        if norm1 == 0:
            for i in range(len(vectors_all)):
                similarities.append(0.0)
            return similarities
        for vector2 in vectors_all:
            norm2 = math.sqrt(VectorUtil._dotProduct(vector2, vector2))
            if norm2 == 0:
                similarities.append(0.0)
            else:
                similarity = VectorUtil._dotProduct(vector1, vector2) / (norm1 * norm2)
                similarities.append(similarity)
        return similarities

    @staticmethod
    def nSimilarity(vector_list1: List[List[float]], vector_list2: List[List[float]]) -> float:
        if len(vector_list1) == 0 or len(vector_list2) == 0:
            raise ValueError("At least one of the passed lists is empty.")
        avg_vector1 = VectorUtil._averageVector(vector_list1)
        avg_vector2 = VectorUtil._averageVector(vector_list2)
        return VectorUtil.similarity(avg_vector1, avg_vector2)

    @staticmethod
    def computeIdfWeightDict(total_num: int, number_dict: Dict[str, float]) -> Dict[str, float]:
        result: Dict[str, float] = {}
        for key, count in number_dict.items():
            weight = math.log((total_num + 1) / (count + 1))
            result[key] = weight
        return result

    @staticmethod
    def _dotProduct(vector1: List[float], vector2: List[float]) -> float:
        result = 0.0
        for i in range(len(vector1)):
            result += vector1[i] * vector2[i]
        return result

    @staticmethod
    def _averageVector(vectors: List[List[float]]) -> List[float]:
        avg_vector = [0.0] * len(vectors[0])
        for vector in vectors:
            for i in range(len(vector)):
                avg_vector[i] += vector[i]
        for i in range(len(avg_vector)):
            avg_vector[i] /= len(vectors)
        return avg_vector